pennylane/transforms/diagonalize_measurements.py

```diff
@@ -0,0 +1,345 @@
+"""Transform to diagonalize measurements on a tape, assuming all measurements are commuting."""
+
+from copy import copy
+from functools import singledispatch
+
+import pennylane as qml
+from pennylane.operation import Tensor
+from pennylane.ops import CompositeOp, LinearCombination, SymbolicOp
+from pennylane.transforms.core import transform
+
+# pylint: disable=protected-access
+
+_default_supported_obs = frozenset(["PauliZ", "Identity"])
+
+
+def null_postprocessing(results):
+    """A postprocessing function returned by a transform that only converts the batch of results
+    into a result for a single ``QuantumTape``.
+    """
+    return results[0]
+
+
+@transform
+def diagonalize_measurements(tape, supported_base_obs=("PauliZ", "Identity")):
+    """Diagonalize the measurements on a tape if they are not supported. Raises an error if the
+    measurements do not commute.
+
+    Args:
+        tape (QNode or QuantumScript or Callable): The quantum circuit to modify the measurements of.
+        supported_base_obs (Optional, Iterable(Str)): A list of names of supported base observables.
+            Allowed names are 'PauliX', 'PauliY', 'PauliZ' and 'Hadamard'. If no list is provided,
+            the transform will diagonalize everything.
```

Comment:
I'm a bit confused by this argument :thinking:

So if you set `supported_base_obs=["X", "Y"]`, then any `X` and `Y` in existing measurements will not be expanded? :thinking: This seems different to diagonalizing, which needs a unique basis.

Reply:
So my concern was for something like:
`supported_base_obs=["X", "Z"]`
with two measurements:
`qml.expval(qml.Z(0)), qml.expval(qml.X(0))`

I wasn't sure what this transform would do :thinking: But it looks like the non-commuting constraint still holds, so I think this is fine.

Reply:
>This seems different to diagonalizing, which needs a unique basis.

I agree that its confusing and not what you normally think of in terms of diagonalization, but the requirements for the shortcut ticket are that we be able to selectively diagonalize only _some_ observables into the Z basis.

For the example you give, it will still raise an error because they don't commute. For `qml.expval(qml.Z(0)), qml.expval(qml.X(1))` and `supported_base_obs=["X", "Z"]` it will return the same thing that went in, since those are both supported.